Use environment variables and Rails.root to remove hardcoded PG SCL references and appliance paths

jrafanie opened the pull request:
https://trello.com/c/aLn9eAj1

Related PRs:
- [ ] https://github.com/ManageIQ/manageiq-appliance/pull/9
- [ ] https://github.com/ManageIQ/manageiq-appliance-build/pull/19

This pull request does the following:
- [x] Externalize the service name and one or more of the paths to a single place PostgresAdmin, which reads the environment variables from https://github.com/ManageIQ/manageiq-appliance/pull/9
- [x] No longer modifies an existing migration
- [x] The configurations table should not reference RPM installation paths, the log collection code should retrieve this from the ENV (in this PR or another) 
- [x] Move/rename MiqPostgresAdmin out of the rails app as PostgresAdmin.

matthewd commented on db/migrate/20131210202928_update_log_collection_path_in_configurations_settings.rb:
```diff
@@ -4,8 +4,8 @@ class Configuration < ActiveRecord::Base
-  NEW_PG_CONF_PATH = "/opt/rh/postgresql92/root/var/lib/pgsql/data/*.conf"
-  NEW_PG_LOGS_PATH = "/opt/rh/postgresql92/root/var/lib/pgsql/data/pg_log/*"
+  NEW_PG_CONF_PATH = "/opt/rh/rh-postgresql94/root/var/lib/pgsql/data/*.conf"
+  NEW_PG_LOGS_PATH = "/opt/rh/rh-postgresql94/root/var/lib/pgsql/data/pg_log/*"
```
Do we really want to change this historical migration?

.. and not make a similar change for currently-up-to-date DBs?

jrafanie replied:
Yes, it would be nice to change the log collection code (the consumer of this configuration setting) to get this information in one place, like a class constant or something, but not in the configuration itself.  I guess I can revisit this when it's closer to ready and either add a new migration to convert postgresql92 to rh-postgresql94 or see about externalizing the setting outside of the configuration itself.  I don't know that it makes sense to even externalize these settings here.

chessbyte replied:
what does the `rh` prefix in `rh-postgresql94` stand for?

jrafanie replied:
Red Hat is my assumption.  I don't know why they changed the convention.  https://www.softwarecollections.org/en/scls/rhscl/rh-postgresql94/

Fryguy replied:
I think it should be a new migration regardless.  Leave the old migration as it was as a snapshot in time.

If we go through a collapse migrations exercise again, this change could be potentially lost.

jrafanie replied:
I agree, it was done as a placeholder/reminder.  I've updated the description of the PR with the TODO list when I return.

matthewd commented:
As a more general observation, it seems like all SCL RPM type stuff should live in the appliance repos... if this repo contains "a normal Rails app", it really oughtn't care about specific paths where other things live. (Though I guess I can give appliance-console a pass.)

Fryguy commented:
Agree with @matthewd ...In fact I was thinking it as soon as we opened this PR...wonder if it can be made generally configurable perhaps in a followup PR.

jrafanie commented:
> Agree with @matthewd ...In fact I was thinking it as soon as we opened this PR...wonder if it can be made generally configurable perhaps in a followup PR.

Yes, I agree.  I wanted to get it opened and verify I could build with it before I was away.
I'd like to externalize as much as possible, probably via one or more environment variables that the appliance/appliance-build repo can set.

I was able to get a build of an appliance with postgres 9.4 to build and all things appear to be operational upon deployment, so I now have a starting point to work with.  At the very least, the migration will have to be fixed in this PR.  Then, I'll see about doing clean up here or in a follow up PR.

jrafanie commented on config/vmdb.tmpl.yml:
```diff
@@ -100,8 +100,6 @@ log:
       - log/*.txt
       - config/*
-      - /opt/rh/postgresql92/root/var/lib/pgsql/data/*.conf
-      - /opt/rh/postgresql92/root/var/lib/pgsql/data/pg_log/*
```
We can drop this from the yml and not migrate data because the log collection code was changed above to add these paths from ENV variables if the configuration didn't already have these paths.

jrafanie commented on app/models/miq_server/log_management.rb:
```diff
@@ -132,6 +134,26 @@ def post_logs(options)
+  end
+
+  def current_log_patterns
+    current_log_pattern_configuration | pg_log_patterns
```
Pull in the current pattern from the configuration and add the pg_log_patterns if the configuration didn't already have them.

chessbyte replied:
Should this be `||` instead of `|`?

jrafanie replied:
No, this is set union.  It's a cheap way to not need a data migration.  The old configuration had the postgresql paths hardcoded:
'/opt/rh/postgresql92/root/var/lib/pgsql/data/*.conf'
'/opt/rh/postgresql92/root/var/lib/pgsql/data/pg_log/*'

This new code reads the existing current pattern array and adds the postgres log patterns from environment variables, and does set union to ensure we only add unique elements in case the current configuration still has the old hardcoded values.  It's not a big deal with they migrate and never remove these values as we won't find logs in old locations so it's benign to have them there.

jrafanie replied:
Clearly, this needs a comment...

jrafanie commented on gems/pending/appliance_console/internal_database_configuration.rb:
```diff
@@ -1,23 +1,24 @@
-    CERT_LOCATION                 = Pathname.new("/var/www/miq/vmdb/certs").freeze
-    POSTGRESQL_SAMPLE             = Pathname.new("/var/www/miq/system/COPY/").join(POSTGRES_DIR).freeze
-    POSTGRESQL_TEMPLATE           = Pathname.new("/var/www/miq/system/TEMPLATE/").join(POSTGRES_DIR).freeze
-    POSTGRESQL_SERVICE            = "postgresql92-postgresql".freeze
```
Moved to PostgresAdmin.

jrafanie commented on spec/lib/vmdb/util_spec.rb:
```diff
@@ -102,7 +102,6 @@ def self.assert_zip_entry_from_path(expected_entry, path)
     assert_zip_entry_from_path("ROOT/www/var/vmdb/miq/log/something.log", "/www/var/vmdb/miq/log/something.log")
     assert_zip_entry_from_path("log/apache/ssl_access.log", "/var/www/miq/vmdb/log/apache/ssl_access.log")
     assert_zip_entry_from_path("config/database.yml", "/var/www/miq/vmdb/config/database.yml")
-    assert_zip_entry_from_path("ROOT/opt/rh/postgresql92/root/var/lib/pgsql/data/pg_hba.conf", "/opt/rh/postgresql92/root/var/lib/pgsql/data/pg_hba.conf")
```
This is an extra test case of a hardcoded PG path we don't need to test here.

jrafanie commented:
@Fryguy @matthewd I think I addressed all of your concerns, please review the description as I tried to update it to reflect the new behavior.  I'll be running a full build to test some of the postgresql areas around running initdb and friends.  Let me know what you think.

jrafanie commented on spec/models/miq_server/update_management_spec.rb:
```diff
@@ -202,6 +203,10 @@
 
   context "#apply_updates" do
+    before do
+      MiqDatabase.stub(:postgres_package_name => "postgresql-server")
```
These stubs seems better than setting whatever the ENV key is that gets populated on appliances.

jrafanie commented on gems/pending/util/postgres_admin.rb:
```diff
@@ -152,11 +186,11 @@ def self.gc(options = {})
     if options[:reindex]
       result = self.reindex(options)
-      _log.info("Output... #{result}") if result.to_s.length > 0
+      $log.info("MIQ(#{self.name}.#{__method__}) Output... #{result}") if result.to_s.length > 0
```
These log lines were changed because we moved the file out of the rails app and don't include Vmdb::Logging anymore.

matthewd commented on gems/pending/appliance_console/internal_database_configuration.rb:
```diff
@@ -1,23 +1,24 @@
+    end
+
+    def self.postgresql_sample
+      Pathname.new("/var/www/miq/system/COPY/").join(postgres_dir)
```
This doesn't seem like it should actually be based on the "real" filesystem location

jrafanie commented:
Thanks @matthewd, made your suggested changes, please see last 4 commits.  I added tests around PostgresAdmin and fixed some bugs those tests found.  The last two commits fix the hardcoded appliance paths you found and one more in PostgresAdmin.

matthewd commented on gems/pending/appliance_console/internal_database_configuration.rb:
```diff
@@ -1,23 +1,24 @@
+    end
+
+    def self.postgresql_sample
+      Rails.root.join("../system/COPY").join(postgres_dir)
```
I mean this should be:

```ruby
Rails.root.join("../system/COPY/opt/rh/postgresql92/root/var/lib/pgsql/data")
```

.. or actually, something more like:

```ruby
Rails.root.join("../system/COPY/pgdata")
```

Either way, the part of the path after `COPY` doesn't depend on where the PG data dir currently lives on the filesystem. [It is a specific well-known path](https://github.com/ManageIQ/manageiq-appliance/tree/06c498e7f890b29b91ac563888ce34ec9402b442/COPY/opt/rh/postgresql92/root/var/lib/pgsql/data). (But because of that, we should also be able to make it a bit less of a mouthful.)

jrafanie replied:
@matthewd The problem is that the system/COPY,LINK directories and the mechanism that copies/symlinks to those locations is dumb.  We blindly copy/link like a chroot below those directories and don't look at what we're copying/linking.  Therefore, the [well-known path](https://github.com/ManageIQ/manageiq-appliance/pull/9/files) is also changing for this work.  The value directory structure below the COPY directory is directly related to the PostgresAdmin.data_directory.

So, we either teach the dumb copier/linker to copy/link to the PostgresAdmin.data_directory or we duplicate this logic by moving files around when we upgrade postgresql.  I went with that I thought was easier to maintain.

jrafanie commented:
@matthewd @gtanzillo This is ready for review/merge.

The other PRs need to be reviewed/merged first:
- [ ] https://github.com/ManageIQ/manageiq-appliance/pull/9
- [ ] https://github.com/ManageIQ/manageiq-appliance-build/pull/19

jrafanie commented:
Doing another test build on the change to use PGDATA env variable.

jrafanie commented:
Finally, ready... test appliance is all good.

Fryguy commented on gems/pending/util/postgres_admin.rb:
```diff
@@ -1,21 +1,57 @@
+  end
+
+  def self.logical_volume_path
+    Pathname.new("/dev").join(VOLUME_GROUP_NAME, LOGICAL_VOLUME_NAME)
```
Ruh Roh

Fryguy replied:
NO TESTS :trollface:
